snapshots/windows/windows.go

```diff
@@ -391,3 +410,154 @@ func (s *snapshotter) parentIDsToParentPaths(parentIDs []string) []string {
 	}
 	return parentLayerPaths
 }
+
+// This is essentially a recreation of what HCS' CreateSandboxLayer does with some extra bells and
+// whistles like expanding the volume if a size is specified. This will create a 1GB scratch
+// vhdx to be used if a different sized scratch that is not equal to the default of 20 is requested.
+func (s *snapshotter) createScratchLayer(ctx context.Context, snDir string, parentLayers []string, sizeGB int) error {
+	parentLen := len(parentLayers)
+	if parentLen == 0 {
+		return errors.New("no parent layers present")
+	}
+	baseLayer := parentLayers[parentLen-1]
+
+	var (
+		templateBase     = filepath.Join(baseLayer, "blank-base.vhdx")
+		templateDiffDisk = filepath.Join(baseLayer, "blank.vhdx")
+		expand           = sizeGB > 0 && sizeGB != 20
+	)
+
+	// If a size greater than 0 or not equal to 20 (the default size produced by the
+	// hcs graphdriver calls) was specified we make a new set of disks to be used.
+	// We make it a 1GB disk and just expand it to the size specified so for future
+	// container runs we don't need to remake a disk.
+	if expand {
+		templateBase = filepath.Join(baseLayer, "scratch.vhdx")
+		templateDiffDisk = filepath.Join(baseLayer, "scratch-diff.vhdx")
+	}
+
+	if _, err := os.Stat(templateDiffDisk); os.IsNotExist(err) {
+		// Scratch disk not present so lets make it.
+		if err := computestorage.SetupContainerBaseLayer(ctx, baseLayer, templateBase, templateDiffDisk, 1); err != nil {
+			return errors.Wrapf(err, "failed to create scratch vhdx at `%s`", baseLayer)
+		}
+	}
+
+	dest := filepath.Join(snDir, "sandbox.vhdx")
+	if err := copyScratchDisk(templateDiffDisk, dest); err != nil {
+		return err
+	}
+
+	if expand {
+		gbToByte := 1024 * 1024 * 1024
+		if err := hcsshim.ExpandSandboxSize(s.info, filepath.Base(snDir), uint64(gbToByte*sizeGB)); err != nil {
+			return errors.Wrapf(err, "failed to expand sandbox vhdx size to %d GB", sizeGB)
+		}
+	}
+
+	return nil
+}
+
+// This handles creating the UVMs scratch layer. This is a bit different from the containers scratch layer in
+// that expanding the volume doesn't work, so a new set of disks will have to be made. Processing the UVM layer also
+// renders any previously made disks unusable. We keep a record of the last requested scratch size and if it differs,
+// we'll make a new set of vhdx's.
+func (s *snapshotter) createUVMScratchLayer(ctx context.Context, snDir string, parentLayers []string, sizeGB int) error {
+	s.scratchLock.Lock()
+	defer s.scratchLock.Unlock()
+
+	// Hardcoded size of the UVMs scratch in the HCS graphdriver calls. If no size is specified
```

Comment:
It is a bit strange to see "graphdriver" here, since there is no graphdrivers in containerd. Do you mean Docker's one?

Reply:
It was related to some internal naming but I'll change this to make more sense.